crates/astria-composer/src/composer.rs

```diff
@@ -59,26 +143,166 @@ impl Composer {
+    /// # Errors
+    /// Returns an error if the listener is not bound
+    pub fn grpc_collector_local_addr(&self) -> io::Result<SocketAddr> {
+        self.grpc_collector.grpc_collector_local_addr()
+    }
+
     /// Runs the composer.
     ///
-    /// Currently only exits if the api server or searcher stop unexpectedly.
-    pub async fn run_until_stopped(self) {
+    /// # Errors
+    /// It errors out if the API Server, Executor or any of the Geth Collectors fail to start.
+    pub async fn run_until_stopped(self) -> eyre::Result<()> {
         let Self {
             api_server,
-            searcher,
+            composer_status_sender,
+            executor,
+            executor_handle,
+            mut geth_collector_tasks,
+            mut geth_collectors,
+            rollups,
+            mut geth_collector_statuses,
+            grpc_collector,
         } = self;
 
-        let api_task =
+        // run the api server
+        let mut api_task =
             tokio::spawn(async move { api_server.await.wrap_err("api server ended unexpectedly") });
-        let searcher_task = tokio::spawn(searcher.run());
 
-        tokio::select! {
-            o = api_task => report_exit("api server", o),
-            o = searcher_task => report_exit("searcher", o),
+        // run the collectors and executor
+        for (chain_id, collector) in geth_collectors.drain() {
+            geth_collector_tasks.spawn(chain_id, collector.run_until_stopped());
+        }
+        let executor_status = executor.subscribe().clone();
+        let mut executor_task = tokio::spawn(executor.run_until_stopped());
+
+        // wait for collectors and executor to come online
+        wait_for_collectors(&geth_collector_statuses).await?;
+        composer_status_sender.send_modify(|status| {
+            status.set_all_collectors_connected(true);
+        });
+        wait_for_executor(executor_status).await?;
+        composer_status_sender.send_modify(|status| {
+            status.set_executor_connected(true);
+        });
+
+        // run the grpc server
+        let mut grpc_server_handle = tokio::spawn(async move {
+            grpc_collector
+                .run_until_stopped()
+                .await
+                .wrap_err("grpc server failed")
+        });
+
+        loop {
+            tokio::select!(
+            o = &mut api_task => {
+                    report_exit("api server unexpectedly ended", o);
+                    return Ok(());
+            },
+            o = &mut executor_task => {
+                    report_exit("executor unexpectedly ended", o);
+                    return Ok(());
+            },
+            o = &mut grpc_server_handle => {
+                    report_exit("grpc server unexpectedly ended", o);
+                    return Ok(());
+            },
+            Some((rollup, collector_exit)) = geth_collector_tasks.join_next() => {
+                reconnect_exited_collector(
```

Comment:
if the geth node goes down for instance, this will continuously loop right? we should probably have a retry limit/backoff here

Reply:
makes sense. we can add it in a seperate PR